### PR TITLE
Tied embeddings in MLP speculator.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Resurrect: https://github.com/huggingface/text-generation-inference/pull/2215/files
Using a single flag to change the layer altogether, means less ifs and less logic everywhere.

Didn't implement scale_input as it's not used by **`ibm-granite/granite-3b-code-instruct-accelerator`** (if it was it would also be a separate layer instead of a if flag.

 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/layers/mlp.py [hunk 1]
```diff
@@ -45,12 +45,95 @@ def forward(self, x):
...
+            weights=weights,
+        )
+
+        # Weights ensure that state_0 accounts for 50% of state magnitude by final head in expectation
+        self.state_weight = 0.5 ** (0.5 / self.n_predict) if self.n_predict > 0 else 1
+        self.emb_weight = math.sqrt(1 - self.state_weight**2)
+        self.activation = nn.GELU()
+        # TODO
+        self.vsize = config.vocab_size
+        self.inner_dim = config.speculator_config["inner_dim"]
```
Comment (OlivierDehaene) [inline server/text_generation_server/layers/mlp.py:81]: Can this get fail?

File: server/text_generation_server/layers/mlp.py [hunk 2]
```diff
@@ -45,12 +45,95 @@ def forward(self, x):
...
+            prefix=f"{prefix}.ln.0",
+            config=config,
+            weights=weights,
+        )
+
+        # Weights ensure that state_0 accounts for 50% of state magnitude by final head in expectation
+        self.state_weight = 0.5 ** (0.5 / self.n_predict) if self.n_predict > 0 else 1
+        self.emb_weight = math.sqrt(1 - self.state_weight**2)
+        self.activation = nn.GELU()
+        # TODO
```
Comment (OlivierDehaene) [inline server/text_generation_server/layers/mlp.py:79]: ?

File: server/text_generation_server/layers/mlp.py [hunk 3]
```diff
@@ -45,12 +45,95 @@ def forward(self, x):
...
+        all_probs = torch.empty(
+            b, self.n_predict, self.vsize, device=state.device
+        )  # b k h v
+        assert (
+            len(top_k_tokens_per_head) == self.n_predict
+        ), f"You must provide a topk number for each head ({self.n_predict} heads, {len(top_k_tokens_per_head)} provided)"
+        for i in range(self.n_predict):
+            # Project and predict
+            z = self.emb(ind)
+            z = z.mul(self.emb_weight * math.sqrt(self.inner_dim / 2))  # b k d
```
Comment (OlivierDehaene) [inline server/text_generation_server/layers/mlp.py:105]: the `self.emb_weight * math.sqrt(self.inner_dim / 2)` constant could be part of the class instead.

Comment (OlivierDehaene) [reply server/text_generation_server/layers/mlp.py:105]: Also I guess we could just multiply the embedding weights by this constant once when creating the class instead of at every forward?

File: server/text_generation_server/layers/mlp.py [hunk 4]
```diff
@@ -45,12 +45,95 @@ def forward(self, x):
...
+            else:
+                state = self.proj1(state) * self.state_weight + z
+            state = self.activation(self.ln(state))  # b k d
+            probs = F.log_softmax(self.head(state), dim=-1)  # b k v
+            _probs, preds = probs.topk(top_k_tokens_per_head[i], dim=-1)  # b k k'
+
+            # Update candidate set with new predictions
+
+            # Update distribution set with new logits
+            all_probs[:, i] = probs.exp()
```
Comment (OlivierDehaene) [inline server/text_generation_server/layers/mlp.py:117]: This is a bit confusing. If they are the logits, why not use the output of `self.head(state)` directly?

File: server/text_generation_server/layers/mlp.py (hunk 1 above)
Comment (Narsil) [reply server/text_generation_server/layers/mlp.py:81]: Hopefully not, but the speculator really needs this value...

File: server/text_generation_server/layers/mlp.py (hunk 2 above)
Comment (Narsil) [reply server/text_generation_server/layers/mlp.py:79]: Removed was there previously.

File: server/text_generation_server/layers/mlp.py (hunk 4 above)
Comment (Narsil) [reply server/text_generation_server/layers/mlp.py:117]: This is the original code from MLP Speculator.
I'm going to keep this here (because there are reasons for its existence in the tree form of the speculation).

**Review by OlivierDehaene — APPROVED** [review]
